listmatch/listmatch.py

```diff
@@ -12,27 +12,26 @@ def __init__(self):
 
 class StateSet(object):
     def __init__(self, states):
-        self.states = set()
-        for state in states:
-            self.add_state(state)
+        self.states = {}
+        for state, paths in states:
+            self.add_state(state, paths)
 
-    def add_state(self, state):
+    def add_state(self, state, paths):
         if state in self.states:
+            self.states[state] += paths
             return
-        self.states.add(state)
+        self.states[state] = paths
         for eps in state.epsilon:
-            self.add_state(eps)
+            self.add_state(eps, paths)
 
     def __iter__(self):
-        return iter(self.states)
+        return iter(self.states.items())
```

Comment:
@adamhadani no, order is unimportant. This was originally a set; just the set of states in which the NFA could be in at the current position in input string. Now it's a dict so that I can associate each state with all the different paths by which we could have arrived at that state